src/Compilers/CSharp/Portable/Symbols/Attributes/SourceAttributeData.cs

```diff
@@ -269,116 +276,155 @@ internal override int GetTargetAttributeSignatureIndex(Symbol targetSymbol, Attr
                             specType = parameterType.GetEnumUnderlyingType()!.SpecialType;
                         }
                     }
-                    else if (parameterType.IsArray())
+                    else if (targetType != (byte)SignatureTypeCode.SZArray && parameterType.IsArray())
                     {
+                        if (targetSignature[signatureByteIndex - 1] != (byte)SignatureTypeCode.SZArray)
+                        {
+                            return false;
```

Comment:
:memo: there was a bug here. This method could not distinguish properly between the `string` case and the `string[]` case. To tighten that logic, I had to refactor a bit (extract a local function).